README.md

# Flutter TestRail

This package provides a Dart interface for seamless integration with the TestRail API, enabling automated test reporting, including test run management and case pass/fail reporting.

## Getting Started

Initialize the TestRail instance using the configure method:

```dart
 FlutterTestRail.initialize(
      username: 'your user name',
      password: 'your password',
      /// The url that points to the test rail server => https://example.testrail.com
      serverDomain: 'https://example.testrail.com',
    );
```



<p><a href="https://buymeacoffee.com/amitsingh6391"> <img align="left" src="https://cdn.buymeacoffee.com/buttons/v2/default-yellow.png" height="50" width="210" alt="amitsingh6391" /></a></p><br><br>


## Usage



### Add Result

Adds a new test result, comment or assigns a test. It’s recommended to use 'Add Result' instead if you plan to add results for multiple tests.

```dart

final testResult = await TestResult.addTestResult(9,statusId:1); //replace 9 with your own testId.

```



### Add Results

Adds one or more new test results, comments, or assigns one or more tests. Ideal for test automation to bulk-add multiple test results in one step.


```dart

List<TestStatus> testStatusResult =[TestStatus(caseId:121,statusId:1,comment:'first test passed'),   TestStatus(caseId:122,statusId:5,comment:'second test failed')]

final testResult = await TestResult.addRunResults(3,addTestResults:testStatusResult); //replace 3 with your own RunId.

```




### Add Results For Cases

Adds one or more new test results, comments or assigns one or more tests (using the case IDs). Ideal for test automation to bulk-add multiple test results in one step.


```dart
/// create your intial Test Stauts  list. 
 List<TestStatus> testStatusResults = [
    TestStatus(
      caseId: 1758,
      statusId: 3,
      comment: 'Verify that user is able to see the splash screen',
    ),
    TestStatus(
      caseId: 1759,
      statusId: 3,
      comment:
          'Verify that after splash screen app user is able to see “Permission screen”',
    ),
  ];

  /// Update your list once your test is completed and add status on test rail with updated status.
 await TestRailUtil.reportMultipleTestCaseResults(testStatusResults);

 class TestRailUtil {
  static void configureTestRail() {
    HttpOverrides.global = MyHttpOverrides();
    TestRail.configure(
      username: 'your user name',
      password: 'your password',
      serverDomain: 'https://example.testrail.com',
    );
  }


  /// Add Result Report to Testrail  [reportSingleTestCaseResult],
  ///
  /// And Equivalent Status Code:
  /// 1: Passed
  /// 2: Blocked
  /// 3: Untested (not allowed when adding a new result)
  /// 4: Retest
  /// 5: Failed
  ///
  ///  So Pass Status Code according to your test status
  static Future<void> reportMultipleTestCaseResults(
    List<TestStatus> testStatusResults,
  ) async {
    final testRun = await TestRun.get(3); //replace 3 with your own Run Id.
    await testRun.addResultsForCases(
      testStatusResults,
    );
  }
}

class MyHttpOverrides extends HttpOverrides {
  @override
  HttpClient createHttpClient(SecurityContext? context) {
    return super.createHttpClient(context)
      ..badCertificateCallback =
          (X509Certificate cert, String host, int port) => true;
  }
}

```

### Create Test case

```dart
/// Create new test case in section
final createdTestCase = await TestCase.create(
  // Replace with your own sectionId
  1,
  title: 'Test case from API',
  customValues: <String, dynamic>{
    // Custom fields start with "custom_" prefix
    'custom_feedback': 'This is custom feedback',
  },
);
```

### Delete Test case
```dart
/// Get TestCase by ID
final testCase = await TestCase.get(1);

await testCase.delete();
```

### Create or Update Runs

```dart
/// Start by creating a new run
final newRun = await TestRun.create(
  name: 'Test execution',
  projectId: 1,
);

/// Add cases to the run
await newRun.updateRun(
  caseIds: [1, 2, 3, 5],
);
```

Once the run is created, results can be reported by case:

```dart
final result = await newRun.addResultForCase(
  caseId: 1,
  statusId: 1,
);

// Optionally add a screenshot or other image to the result
await result.addAttachment(
  '/workspace/attachments/failure.png',
);
```

### Get

Historical runs, cases, and sections can be retrieved:

```dart
final testCase = await TestCase.get(1);

final testCases = await TestCase.getAll(1);

final testRun = await TestRun.get(1);

final testSection = await TestSection.get(1);

final testCaseHistory = await TestCaseHistory.get(1);
```

Completed or ongoing test run results can be retrieved:

```dart
final caseResults = await TestResult.getCaseResults(
  // Case ID is from TestCases, not TestRun
  184234,
  runId: 1833,
);

final runResults = await TestResult.getRunResults(
  1818,
  statusId: [5, 1],
);

final testResults = await TestResult.getTestResults(
  // Test ID from particular TestRun
  1868150,
);
```

CHANGELOG.md
## 1.0.3
Add new method getReports and genrateReport to get test reports on Mail.
## 1.0.2
Updated dependencies

## 1.0.1
Update Readme file.

## 1.0.0

********** BREAKING CHANGES *********

1 . TestRail.configure() is now FlutterTestRail.initialize()

2. Add New Method to add Single test Result by : TestResult.addTestResult(9,statusId:1)

3. Now you can also add multiple Test results in one time :
  ### TestResult.addRunResults(3,addTestResults:testStatusResult);

## 0.0.4
Add Support for latest dart sdk: '>=3.0.1 <4.0.0'
## 0.0.3
Update Project to Support Dart sdk: '>=2.19.0 <3.0.0'

## 0.0.2
Update ReadMe

## 0.0.1

* Initial.


lib/test_run.dart
import 'flutter_testrail.dart';

class TestRun {
  final int? assignedtoId;
  final int? blockedCount;
  final DateTime? completedOn;
  final String? config;
  final List<int>? configIds;
  final int? createdBy;
  final DateTime? createdOn;
  final int? customStatus1Count;
  final int? customStatus2Count;
  final int? customStatus3Count;
  final int? customStatus4Count;
  final int? customStatus5Count;
  final int? customStatus6Count;
  final int? customStatus7Count;
  final String? description;
  final int? failedCount;
  final int id;
  final bool? includeAll;
  final bool? isCompleted;
  final int? milestoneId;
  final String? name;
  final int? passedCount;
  final int? planId;
  final int? projectId;
  final String? refs;
  final int? retestCount;
  final int? suiteId;
  final int? untestedCount;
  final DateTime? updatedOn;
  final String? url;

  TestRun({
    this.assignedtoId,
    this.blockedCount,
    this.completedOn,
    this.config,
    this.configIds,
    this.createdBy,
    this.createdOn,
    this.customStatus1Count,
    this.customStatus2Count,
    this.customStatus3Count,
    this.customStatus4Count,
    this.customStatus5Count,
    this.customStatus6Count,
    this.customStatus7Count,
    this.description,
    this.failedCount,
    required this.id,
    this.includeAll,
    this.isCompleted,
    this.milestoneId,
    this.name,
    this.passedCount,
    this.planId,
    this.projectId,
    this.refs,
    this.retestCount,
    this.suiteId,
    this.untestedCount,
    this.updatedOn,
    this.url,
  });

  factory TestRun.fromJson(Map<String, dynamic> json) {
    DateTime? completedOn;
    DateTime? createdOn;
    DateTime? updatedOn;

    if (json['completed_on'] != null) {
      completedOn = DateTime.fromMillisecondsSinceEpoch(json['completed_on']);
    }

    if (json['created_on'] != null) {
      createdOn = DateTime.fromMillisecondsSinceEpoch(json['created_on']);
    }

    if (json['updated_on'] != null) {
      updatedOn = DateTime.fromMillisecondsSinceEpoch(json['updated_on']);
    }

    return TestRun(
      assignedtoId: json['assignedto_id'],
      blockedCount: json['blocked_count'],
      completedOn: completedOn,
      config: json['config'],
      configIds: json['config_ids']?.cast<int>(),
      createdBy: json['created_by'],
      createdOn: createdOn,
      customStatus1Count: json['custom_status1_count'],
      customStatus2Count: json['custom_status2_count'],
      customStatus3Count: json['custom_status3_count'],
      customStatus4Count: json['custom_status4_count'],
      customStatus5Count: json['custom_status5_count'],
      customStatus6Count: json['custom_status6_count'],
      customStatus7Count: json['custom_status7_count'],
      description: json['description'],
      failedCount: json['failed_count'],
      id: json['id'],
      includeAll: json['include_all'],
      isCompleted: json['is_completed'],
      milestoneId: json['milestone_id'],
      name: json['name'],
      passedCount: json['passed_count'],
      planId: json['plan_id'],
      projectId: json['project_id'],
      refs: json['refs'],
      retestCount: json['retest_count'],
      suiteId: json['suite_id'],
      untestedCount: json['untested_count'],
      updatedOn: updatedOn,
      url: json['url'],
    );
  }

  Future<TestResult> addResultForCase(
    int caseId, {
    String? comment,
    String? elapsed,
    required int statusId,
  }) async {
    final response = await FlutterTestRail.instance.client.request(
      '/add_result_for_case/$id/$caseId',
      RequestMethod.post,
      params: {
        'status_id': statusId,
        if (comment != null) 'comment': comment,
        if (elapsed != null) 'elapsed': elapsed,
      },
    );

    return TestResult.fromJson(response!);
  }

  Future<void> addResultsForCases(
    List<TestStatus> testStatusResults,
  ) async {
    final results = testStatusResults.map((status) => status.toJson()).toList();
    await FlutterTestRail.instance.client.request(
      '/add_results_for_cases/$id',
      RequestMethod.post,
      params: {
        'results': results,
      },
    );
  }

  Future<TestRun> close() async {
    final response = await FlutterTestRail.instance.client
        .request('/close_run/$id', RequestMethod.post);
    return TestRun.fromJson(response!);
  }

  static Future<TestRun> create({
    Iterable<int>? caseIds,
    String? description,
    bool includeAll = false,
    required String name,
    required int projectId,
  }) async {
    final response = await FlutterTestRail.instance.client.request(
      '/add_run/$projectId',
      RequestMethod.post,
      params: <String, dynamic>{
        if (caseIds != null) 'case_ids': caseIds,
        if (description != null) 'description': description,
        'include_all': includeAll,
        'name': name,
      },
    );
    return TestRun.fromJson(response!);
  }

  static Future<TestRun> get(int runId) async {
    final response = await FlutterTestRail.instance.client
        .request('/get_run/$runId', RequestMethod.get);
    return TestRun.fromJson(response!);
  }

  static Future<TestRuns> getAll({
    required int projectId,
    DateTime? createdAfter,
    DateTime? createdBefore,
    Iterable<int>? createdBy,
    bool? isCompleted,
    int? limit,
    int? offset,
    Iterable<int>? milestoneId,
    String? refsFilter,
    Iterable<int>? suiteId,
  }) async {
    final queryParameters = <String, dynamic>{
      'created_after': createdAfter?.millisecondsSinceEpoch,
      'created_before': createdBefore?.millisecondsSinceEpoch,
      'created_by': createdBy?.join(','),
      'is_completed': isCompleted,
      'limit': limit,
      'offset': offset,
      'milestone_id': milestoneId?.join(','),
      'refs_filter': refsFilter,
      'suite_id': suiteId?.join(','),
    };

    queryParameters.removeWhere((_, dynamic value) => value == null);

    final response = await FlutterTestRail.instance.client.request(
      '/get_runs/$projectId',
      RequestMethod.get,
      queryParameters: queryParameters,
    );

    return TestRuns.fromJson(response!);
  }

  Future<TestRun> update({
    required Iterable<int> caseIds,
    bool includeAll = false,
  }) async {
    final response = await FlutterTestRail.instance.client.request(
      '/update_run/$id',
      RequestMethod.post,
      params: {
        'case_ids': caseIds,
        'include_all': includeAll,
      },
    );
    return TestRun.fromJson(response!);
  }

  Map<String, dynamic> get asJson => {
        'assignedto_id': assignedtoId,
        'blocked_count': blockedCount,
        'completed_on': completedOn?.millisecondsSinceEpoch,
        'config_ids': configIds,
        'config': config,
        'created_by': createdBy,
        'created_on': createdOn?.millisecondsSinceEpoch,
        'custom_status1_count': customStatus1Count,
        'custom_status2_count': customStatus2Count,
        'custom_status3_count': customStatus3Count,
        'custom_status4_count': customStatus4Count,
        'custom_status5_count': customStatus5Count,
        'custom_status6_count': customStatus6Count,
        'custom_status7_count': customStatus7Count,
        'description': description,
        'failed_count': failedCount,
        'id': id,
        'include_all': includeAll,
        'is_completed': isCompleted,
        'milestone_id': milestoneId,
        'name': name,
        'passed_count': passedCount,
        'plan_id': planId,
        'project_id': projectId,
        'refs': refs,
        'retest_count': retestCount,
        'suite_id': suiteId,
        'untested_count': untestedCount,
        'updated_on': updatedOn?.millisecondsSinceEpoch,
        'url': url,
      };
}

class TestStatus {
  TestStatus({
    required this.caseId,
    required this.statusId,
    this.comment,
    this.elapsed,
    this.version,
    this.defects,
  });
  final int caseId;
  final int statusId;
  final String? comment;
  final String? elapsed;
  final String? version;
  final String? defects;

  Map<String, dynamic> toJson() {
    return {
      'case_id': caseId,
      'status_id': statusId,
      'comment': comment ?? 'comment',
      'elapsed': elapsed ?? '5m',
      'version': version ?? '1.0 RC1',
      'defects': defects ?? 'TR-7',
    };
  }
}


lib/test_case.dart
import 'flutter_testrail.dart';

class TestCase {
  final int? createdBy;
  final DateTime? createdOn;
  final Map<String, dynamic>? customAttributes;
  final int? displayOrder;
  final String? estimate;
  final String? estimateForecast;
  final int id;
  final int? isDeleted;
  final int? milestoneId;
  final int? priorityId;
  final String? refs;
  final int? sectionId;
  final int? suiteId;
  final int? templateId;
  final String? title;
  final int? typeId;
  final int? updatedBy;
  final DateTime? updatedOn;

  TestCase({
    this.createdBy,
    this.createdOn,
    this.customAttributes,
    this.displayOrder,
    this.estimate,
    this.estimateForecast,
    required this.id,
    this.isDeleted,
    this.milestoneId,
    this.priorityId,
    this.refs,
    this.sectionId,
    this.suiteId,
    this.templateId,
    this.title,
    this.typeId,
    this.updatedBy,
    this.updatedOn,
  });

  factory TestCase.fromJson(Map<String, dynamic> json) {
    DateTime? createdOn;
    DateTime? updatedOn;

    if (json['created_on'] != null) {
      createdOn = DateTime.fromMillisecondsSinceEpoch(json['created_on']);
    }

    if (json['updated_on'] != null) {
      updatedOn = DateTime.fromMillisecondsSinceEpoch(json['updated_on']);
    }

    Map<String, dynamic> customAttributes = {};
    var customStepsSeparated = [];

    for (final element in json.entries) {
      if (element.key.contains("custom_")) {
        customAttributes.putIfAbsent(element.key, () => element.value);
      }

      if (element.key == "custom_steps_separated" && element.value != null) {
        customStepsSeparated = List.from(
          element.value.map(
            (stepJson) => TestCaseCustomStep.fromJson(stepJson),
          ),
        );
        customAttributes.update(element.key, (value) => customStepsSeparated);
      }
    }

    return TestCase(
      createdBy: json['created_by'],
      createdOn: createdOn,
      customAttributes: customAttributes['custom_attributes'],
      displayOrder: json['display_order'],
      estimate: json['estimate'],
      estimateForecast: json['estimate_forecast'],
      id: json['id'],
      isDeleted: json['is_deleted'],
      milestoneId: json['milestone_id'],
      priorityId: json['priority_id'],
      refs: json['refs'],
      sectionId: json['section_id'],
      suiteId: json['suite_id'],
      templateId: json['template_id'],
      title: json['title'],
      typeId: json['type_id'],
      updatedBy: json['updated_by'],
      updatedOn: updatedOn,
    );
  }

  static Future<TestCase> create(
    int sectionId, {
    Map<String, dynamic>? customValues,
    String? estimate,
    int? milestoneId,
    int? priorityId,
    String? refs,
    required String title,
    int? templateId,
    int? typeId,
  }) async {
    final parametersMap = {
      if (estimate != null) 'estimate': estimate,
      if (milestoneId != null) 'milestone_id': milestoneId,
      if (priorityId != null) 'priority_id': priorityId,
      if (refs != null) 'refs': refs,
      if (templateId != null) 'template_id': templateId,
      'title': title,
      if (typeId != null) 'type_id': typeId,
    };

    customValues?.forEach((key, value) {
      parametersMap.putIfAbsent(key, () => value);
    });

    final response = await FlutterTestRail.instance.client.request(
      '/add_case/$sectionId',
      RequestMethod.post,
      params: parametersMap,
    );

    return TestCase.fromJson(response!);
  }

  Future<SoftDeleteTestCase?> delete({
    bool soft = false,
  }) async {
    final parametersMap = {
      'soft': soft ? 1 : 0,
    };
    final url = '/delete_case/$id';
    final response = await FlutterTestRail.instance.client.request(
      url,
      RequestMethod.post,
      params: parametersMap,
    );

    if (soft) {
      return SoftDeleteTestCase.fromJson(response!);
    }

    return null;
  }

  static Future<TestCase> get(int caseId) async {
    final response = await FlutterTestRail.instance.client
        .request('/get_case/$caseId', RequestMethod.get);
    return TestCase.fromJson(response!);
  }

  static Future<TestCases> getAll(
    int projectId, {
    DateTime? createdAfter,
    DateTime? createdBefore,
    Iterable<int>? createdBy,
    String? filter,
    int? limit,
    int? offset,
    Iterable<int>? milestoneId,
    Iterable<int>? priorityId,
    String? refs,
    int? sectionId,
    Iterable<int>? templateId,
    Iterable<int>? typeId,
    DateTime? updatedAfter,
    DateTime? updatedBefore,
    Iterable<int>? updatedBy,
  }) async {
    final queryParameters = <String, dynamic>{
      'created_after': createdAfter?.millisecondsSinceEpoch,
      'created_before': createdBefore?.millisecondsSinceEpoch,
      'created_by': createdBy?.join(','),
      'filter': filter,
      'limit': limit,
      'offset': offset,
      'milestone_id': milestoneId?.join(','),
      'priority_id': priorityId?.join(','),
      'refs_filter': refs,
      'section_id': sectionId,
      'template_id': templateId?.join(','),
      'type_id': typeId?.join(','),
      'updated_after': updatedAfter?.millisecondsSinceEpoch,
      'updated_before': updatedBefore?.millisecondsSinceEpoch,
      'updated_by': updatedBy?.join(','),
    };

    queryParameters.removeWhere((_, dynamic value) => value == null);

    final response = await FlutterTestRail.instance.client.request(
      '/get_cases/$projectId',
      RequestMethod.get,
      queryParameters: queryParameters,
    );
    return TestCases.fromJson(response!);
  }

  Map<String, dynamic> get asJson => {
        'created_by': createdBy,
        'created_on': createdOn?.millisecondsSinceEpoch,
        'custom_attributes': customAttributes,
        'display_order': displayOrder,
        'estimate': estimate,
        'estimate_forecast': estimateForecast,
        'id': id,
        'is_deleted': isDeleted,
        'milestone_id': milestoneId,
        'priority_id': priorityId,
        'refs': refs,
        'section_id': sectionId,
        'suite_id': suiteId,
        'template_id': templateId,
        'title': title,
        'type_id': typeId,
        'updated_by': updatedBy,
        'updated_on': updatedOn?.millisecondsSinceEpoch,
      };
}

class TestCaseCustomStep {
  final String? additionalInfo;
  final String? content;
  final String? expected;
  final String? refs;
  final int? sharedStepId;

  TestCaseCustomStep({
    this.additionalInfo,
    this.content,
    this.expected,
    this.refs,
    this.sharedStepId,
  });

  factory TestCaseCustomStep.fromJson(Map<String, dynamic> json) {
    return TestCaseCustomStep(
      additionalInfo: json['additional_info'],
      content: json['content'],
      expected: json['expected'],
      refs: json['refs'],
      sharedStepId: json['shared_step_id'],
    );
  }

  Map<String, dynamic> get asJson => {
        'additional_info': additionalInfo,
        'content': content,
        'expected': expected,
        'refs': refs,
        if (sharedStepId != null) 'shared_step_id': sharedStepId,
      };
}

class SoftDeleteTestCase {
  final int? results;
  final int? tests;

  SoftDeleteTestCase({
    this.results,
    this.tests,
  });

  factory SoftDeleteTestCase.fromJson(Map<String, dynamic> json) {
    return SoftDeleteTestCase(tests: json['tests'], results: json['results']);
  }

  Map<String, dynamic> get asJson => {
        'results': results,
        'tests': tests,
      };
}


lib/test_rail.dart
import 'package:http/http.dart' as http;
import 'flutter_testrail.dart';

class FlutterTestRail {
  FlutterTestRail._({
    required this.client,
  });
  final FlutterTestRailHttpClient client;

  static late FlutterTestRail instance;

  static FlutterTestRail initialize({
    required String password,

    /// The url that points to the TestRail server => https://example.testrail.com
    required String serverDomain,
    required String username,
    http.Client? client,
  }) {
    return instance = FlutterTestRail._(
      client: FlutterTestRailHttpClient(
        client: client,
        password: password,
        serverDomain: serverDomain,
        username: username,
      ),
    );
  }
}


lib/src/flutter_test_rail_http_client.dart
import 'dart:convert';
import 'package:http/http.dart' as http;

enum RequestMethod { get, post, postMultipart }

class FlutterTestRailHttpClient {
  final http.Client client;
  final String password;

  /// The url that points to the TestRail server => https://example.testrail.com
  final String serverDomain;
  final String username;

  static const String _urlExtension = '/index.php?';
  static const String _apiVersion = '/api/v2';

  FlutterTestRailHttpClient({
    http.Client? client,
    required this.password,
    required this.serverDomain,
    required this.username,
  }) : client = client ?? http.Client();

  Future<Map<String, dynamic>?> request(
    String endpoint,
    RequestMethod method, {
    Map<String, dynamic>? queryParameters,
    // Include `filePath` for `RequestMethod.postMultipart` requests
    String? filePath,
    Map<String, dynamic> params = const {},
  }) async {
    final url = '$serverDomain.$_urlExtension$_apiVersion/$endpoint';
    final authorization =
        'Basic ${base64Encode(utf8.encode('$username:$password'))}';
    final header = {'authorization': authorization};

    http.Response response;

    switch (method) {
      case RequestMethod.get:
        final protocol = serverDomain.contains('https://') ? 'https' : 'http';

        /// Extract only domain
        final simpleDomain = serverDomain.replaceAll(protocol, '').substring(3);

        Map<String, String> filterParams = {'$_apiVersion$endpoint': ''};

        queryParameters?.forEach((key, value) {
          filterParams.putIfAbsent(key, () => value.toString());
        });

        final requestUrl = Uri(
          scheme: protocol,
          host: simpleDomain,
          path: 'index.php',
          queryParameters: filterParams,
        );

        response = await client.get(requestUrl, headers: header);
        break;
      case RequestMethod.post:
        final headers = {...header, 'Content-Type': 'application/json'};
        response = await client.post(Uri.parse(url),
            headers: headers, body: jsonEncode(params));
        break;
      case RequestMethod.postMultipart:
        final headers = {...header, 'Content-Type': 'multipart/form-data'};

        final request = http.MultipartRequest('POST', Uri.parse(url));
        final file = await http.MultipartFile.fromPath('attachment', filePath!);

        request.files.add(file);
        request.headers.addAll(headers);

        final streamedResponse = await request.send();
        response = await http.Response.fromStream(streamedResponse);
        break;
    }
    final decodedBodyIsList = jsonDecode(response.body) is List<dynamic>;
    if (response.body.isNotEmpty && !decodedBodyIsList) {
      return jsonDecode(response.body) as Map<String, dynamic>;
    } else {
      return {'reports': jsonDecode(response.body)};
    }
  }
}


example/main.dart
import 'dart:io';
import 'package:flutter_testrail/flutter_testrail.dart';
import 'package:flutter_test/flutter_test.dart';

import 'myapp.dart';

Future<void> main() async {
  setUpAll(() async {
    TestRailUtil.configureTestRail();
  });

  List<TestStatus> testStatusResults = [
    TestStatus(
      caseId: 1758,
      statusId: 3,
      comment: 'Verify that user is able to see the splash screen',
    ),
    TestStatus(
      caseId: 1759,
      statusId: 3,
      comment:
          'Verify that after splash screen app user is able to see “Permission screen”',
    ),
  ];

  tearDownAll(() async {
    ///  *At the end we will update our test status on TestRail.
    await TestRailUtil.reportMultipleTestCaseResults(testStatusResults);
  });

  group('Test Permission flow', () {
    testWidgets(
      'Grant permissions',
      (tester) async {
        testStatusResults = TestRailUtil.getUpdatdTestStatusList(
          caseId: 1759,
          statusId: 5,
          testStatusResults: testStatusResults,
        );

        /// * Pump Your widget here.
        tester.pumpWidget(MyApp());

        await tester.pumpAndSettle(const Duration(seconds: 5));

        /// * Find you expected Text here.
        expect(find.text('text'), findsOneWidget);

        await tester.pumpAndSettle();

        /// * Once Your test is passed update your intial test list.
        testStatusResults = TestRailUtil.getUpdatdTestStatusList(
          caseId: 1761,
          statusId: 1,
          testStatusResults: testStatusResults,
        );
      },
    );
  });
}

class TestRailUtil {
  static void configureTestRail() {
    HttpOverrides.global = MyHttpOverrides();
    FlutterTestRail.initialize(
      username: 'your user name',
      password: 'your password',
      serverDomain: 'https://example.testrail.com',
    );
  }

  /// Add Result Report to Testrail  [reportSingleTestCaseResult],
  ///
  /// And Equivalent Status Code:
  /// 1: Passed
  /// 2: Blocked
  /// 3: Untested (not allowed when adding a new result)
  /// 4: Retest
  /// 5: Failed
  ///
  ///  * So Pass Status Code according to your test status
  static Future<void> reportSingleTestCaseResult({
    required int testCaseId,
    required int testStatusId,
    String? comment,
    String? elapsed,
  }) async {
    final testRun = await TestRun.get(3);
    await testRun.addResultForCase(
      testCaseId,
      statusId: testStatusId,
      comment: comment ?? 'comment',
      elapsed: elapsed,
    );
  }

  /// Add Result Report to Testrail  [reportSingleTestCaseResult],
  ///
  /// And Equivalent Status Code:
  /// 1: Passed
  /// 2: Blocked
  /// 3: Untested (not allowed when adding a new result)
  /// 4: Retest
  /// 5: Failed
  ///
  ///  * So Pass Status Code according to your test status
  static Future<void> reportMultipleTestCaseResults(
    List<TestStatus> testStatusResults,
  ) async {
    final testRun = await TestRun.get(3);
    await testRun.addResultsForCases(
      testStatusResults,
    );
  }

  static List<TestStatus> getUpdatdTestStatusList({
    required List<TestStatus> testStatusResults,
    required int caseId,
    required int statusId,
  }) {
    final updatedTestStatusResults = testStatusResults.map((teststatus) {
      if (teststatus.caseId == caseId) {
        return TestStatus(
          caseId: teststatus.caseId,
          statusId: statusId,
          comment: teststatus.comment,
        );
      }
      return teststatus;
    }).toList();

    return updatedTestStatusResults;
  }
}

class MyHttpOverrides extends HttpOverrides {
  @override
  HttpClient createHttpClient(SecurityContext? context) {
    return super.createHttpClient(context)
      ..badCertificateCallback =
          (X509Certificate cert, String host, int port) => true;
  }
}
